<a href="https://colab.research.google.com/github/shrivatsashetty/Projects/blob/main/Synamedia_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Import modules**

In [81]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
# %matplotlib notebook
!pip install pandas-profiling

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 324.4/324.4 kB 4.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 357.3/357.3 kB 10.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 102.7/102.7 kB 7.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 679.5/679.5 kB 11.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 296.5/296.5 kB 14.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.7/4.7 MB 31.0 MB/s eta 0:00:00
  Created wheel for htmlmin: filename=htmlmin-0.1.12-py3-none-any.whl size=27081 sha256=6dd434e45c69813aa36270cf9cec505a162e086757c5f68805c6fd87934714f5
  Stored in directory: /root/.cache/pip/wheels/dd/91/29/a79cecb328d01739e64017b6fb9a1ab9d8cb1853098ec5966d
Successfully built htmlmin


In [2]:
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [3]:
main_df = pd.read_csv("/content/drive/MyDrive/Data-Science/2DaysViewingSession.csv")

# **Tabular form**

In [4]:
main_df

state                          event_start_tz  \
0             Selangor  2023-08-27 11:03:13.686 Asia/Singapore   
1         Kuala Lumpur  2023-08-27 10:48:56.982 Asia/Singapore   
2                Johor  2023-08-27 10:56:51.851 Asia/Singapore   
3         Kuala Lumpur  2023-08-27 10:31:40.717 Asia/Singapore   
4               Penang  2023-08-27 10:24:16.194 Asia/Singapore   
...                ...                                     ...   
26011962  Kuala Lumpur  2023-08-28 15:04:59.497 Asia/Singapore   
26011963  Kuala Lumpur  2023-08-28 15:05:03.023 Asia/Singapore   
26011964  Kuala Lumpur  2023-08-28 15:05:16.592 Asia/Singapore   
26011965       Sarawak  2023-08-28 14:59:50.687 Asia/Singapore   
26011966  Kuala Lumpur  2023-08-28 15:00:12.829 Asia/Singapore   

          duration_seconds content_type                      content  \
0                       40          VOD                  Ombak Rindu   
1                      424          VOD    Pada Zaman Dahulu S4 Ep08   
2                      398          VOD         Upin & Ipin S16 Ep34   
3                     1411          VOD         Tanah Kubur S13 Ep08   
4                     2526          VOD        Narcotics Heroes Ep17   
...                    ...          ...                          ...   
26011962                 1         CDVR          Ghost of Relativity   
26011963               906         CDVR          Ghost of Relativity   
26011964               362          VOD         Upin & Ipin S15 Ep34   
26011965              1241          VOD               Mechamato Ep01   
26011966               672          VOD  Tayo The Little Bus S5 Ep13   

          content_source  
0               astro.2d  
1               astro.7d  
2              astro.30d  
3              astro.30d  
4              astro.30d  
...                  ...  
26011962  TVB CLASSIC HD  
26011963  TVB CLASSIC HD  
26011964       astro.30d  
26011965       astro.30d  
26011966        astro.7d  

[26011967 rows x 6 columns]

# *main dataframe shape*

In [5]:
main_df.shape

(26011967, 6)

### ***the main dataframe contains 2,60,11,967 records***

# main dataset columns

In [6]:
i = 1
print("The dataset contains the following columns")
for column in main_df.columns:
  print(i,".",column)
  i+=1

The dataset contains the following columns
1 . state
2 . event_start_tz
3 . duration_seconds
4 . content_type
5 . content
6 . content_source


In [7]:
main_df.describe()

duration_seconds
count      2.601197e+07
mean       7.747659e+02
std        1.274039e+03
min        0.000000e+00
25%        5.700000e+01
50%        2.490000e+02
75%        9.730000e+02
max        3.239500e+04

In [8]:
main_df.isna().sum()

state               13493121
event_start_tz             0
duration_seconds           0
content_type               0
content               110136
content_source          3780
dtype: int64

In [9]:
cleaned_df = main_df.dropna()

In [10]:
cleaned_df.shape

(12466618, 6)

In [11]:
type(cleaned_df.shape)

tuple

# difference b/w new and orignal data frame

In [12]:
records_differenece = main_df.shape[0] - cleaned_df.shape[0]

In [13]:
percent_record_difference = (records_differenece / main_df.shape[0]) * 100

### percentage difference in records after dropping rows

In [14]:
percent_record_difference

52.073528311027

In [15]:
cleaned_df.describe()

duration_seconds
count      1.246662e+07
mean       5.246197e+02
std        1.160204e+03
min        0.000000e+00
25%        1.000000e+01
50%        1.250000e+02
75%        4.280000e+02
max        3.239500e+04

In [16]:
cleaned_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12466618 entries, 0 to 26011966
Data columns (total 6 columns):
 #   Column            Dtype 
---  ------            ----- 
 0   state             object
 1   event_start_tz    object
 2   duration_seconds  int64 
 3   content_type      object
 4   content           object
 5   content_source    object
dtypes: int64(1), object(5)
memory usage: 665.8+ MB


In [17]:
uniq_states = cleaned_df['state'].value_counts()

### List of distinct states

In [18]:
uniq_states


Kuala Lumpur                  6878719
Selangor                      2634243
Penang                         751319
Johor                          722522
Perak                          404518
Sarawak                        228747
Negeri Sembilan                219438
Sabah                          208386
Melaka                         119456
Kedah                           94178
Kelantan                        62062
Pahang                          52238
UNKNOWN                         48884
Terengganu                      31681
Putrajaya                        7271
Perlis                           1869
Karnataka                         379
Labuan                            370
ERROR                             169
Maharashtra                       129
Seoul                              10
Uttar Pradesh                       6
Occitanie                           5
England                             5
Victoria                            3
Jerusalem                           3
Tamil Nadu  

# List of actual Malaysian states plus federal territories

In [19]:
states_list = ["Perlis", "Kedah", "Penang", "Perak", "Selangor", "Negeri Sembilan",
               "Melaka", "Johor", "Kelantan", "Terengganu", "Pahang", "Kuala Lumpur", "Putrajaya", "Sabah", "Sarawak", "Labuan" ]

In [20]:
len(states_list)

16

# filtering duplicate state if any

In [21]:
uniq_states_list = list(set(states_list))

In [22]:
for state in uniq_states_list:
  print(state)

Penang
Perlis
Labuan
Selangor
Kuala Lumpur
Sarawak
Sabah
Kelantan
Pahang
Melaka
Putrajaya
Terengganu
Negeri Sembilan
Johor
Kedah
Perak


In [23]:
len(uniq_states_list)

16

In [24]:
print("Total distinct states:",uniq_states.count())

Total distinct states: 31


In [25]:
org_distinct_states = main_df['state'].value_counts()

In [26]:
org_distinct_states.count()

31

# Filtering  non Malaysian states

In [27]:
cleaned_df = cleaned_df[cleaned_df['state'].isin(uniq_states_list)]

In [28]:
type(cleaned_df['state'].isin(uniq_states_list))

pandas.core.series.Series

In [29]:
cleaned_df_states = cleaned_df['state'].value_counts()

In [30]:
cleaned_df_states

Kuala Lumpur       6878719
Selangor           2634243
Penang              751319
Johor               722522
Perak               404518
Sarawak             228747
Negeri Sembilan     219438
Sabah               208386
Melaka              119456
Kedah                94178
Kelantan             62062
Pahang               52238
Terengganu           31681
Putrajaya             7271
Perlis                1869
Labuan                 370
Name: state, dtype: int64

In [31]:
type(cleaned_df_states)

pandas.core.series.Series

In [32]:
cleaned_df_states.sum()

12417017

In [33]:
cleaned_df_states.count()

16

### hence the **final version of cleaned_df** contains list of only 16 states which is actual no of states in malaysia

In [34]:
cleaned_df.shape

(12417017, 6)

## Till here we have a final version of cleaned dataframe

# Now scaling down the dataset

In [55]:
sample_size = int(cleaned_df.shape[0] * 0.4)

In [56]:
sample_size

4966806

In [57]:
sampled_df = cleaned_df.sample(n=sample_size, random_state=42)

In [58]:
sampled_states_series = sampled_df["state"].value_counts()

In [59]:
sampled_states_series

Kuala Lumpur       2752122
Selangor           1052497
Penang              301065
Johor               289287
Perak               161315
Sarawak              91277
Negeri Sembilan      87627
Sabah                83336
Melaka               47890
Kedah                37831
Kelantan             24785
Pahang               21128
Terengganu           12850
Putrajaya             2883
Perlis                 755
Labuan                 158
Name: state, dtype: int64

In [60]:
cumilative_frequency_of_states = sampled_states_series.sum()

In [61]:
cumilative_frequency_of_states

4966806

In [65]:
statewise_frequency_percentages = {}
for state in sampled_states_series.index:
  statewise_frequency_percentages[state] = (sampled_states_series[state] / cumilative_frequency_of_states) * 100

In [66]:
statewise_frequency_percentages

{'Kuala Lumpur': 55.410297885602944,
 'Selangor': 21.190620289981126,
 'Penang': 6.061541360785986,
 'Johor': 5.824407073680752,
 'Perak': 3.247861905619023,
 'Sarawak': 1.8377403909071544,
 'Negeri Sembilan': 1.7642525196273018,
 'Sabah': 1.677858970130905,
 'Melaka': 0.964201138518396,
 'Kedah': 0.7616766187364676,
 'Kelantan': 0.4990128464852463,
 'Pahang': 0.4253840395618431,
 'Terengganu': 0.25871757423181013,
 'Putrajaya': 0.058045351479401444,
 'Perlis': 0.015200915840079116,
 'Labuan': 0.003181118811566226}

# list of distinct content_type (content_category)


In [67]:
content_categories = sampled_df['content_type'].value_counts()

In [68]:
content_categories

VOD                2933893
LINEAR              884036
CDVR                693233
Live Pause          183495
Play From Start     160127
IP Switch           112022
Name: content_type, dtype: int64

# list of distinct content (T.V Shows)

In [70]:
content_series = sampled_df['content'].value_counts()

 * ***the value_counts() method will always return a pandas series with unique indexes so no need to worry about the uniqueness of index names***

In [71]:
content_series

(L) BWF World Championships: Finals     71166
(L) PL: Newcastle vs Liverpool          65977
(L) PL: Sheffield vs Manchester City    44857
The Flash [Fast Premiere]               41814
Tiktok Awards Malaysia                  32574
                                        ...  
Indira Ep184                                1
Little Food Hunter S1 Ep07                  1
My Little Monkey King S1 Ep14               1
Room 104 S1 Ep09                            1
Tong Tong's Wonderland S3 Ep10              1
Name: content, Length: 55068, dtype: int64

In [80]:
len(content_series)

55068

## There are 55068 distinct shows below is the top 10 most occuring TV shows


In [77]:
content_series.head(10)

(L) BWF World Championships: Finals       71166
(L) PL: Newcastle vs Liverpool            65977
(L) PL: Sheffield vs Manchester City      44857
The Flash [Fast Premiere]                 41814
Tiktok Awards Malaysia                    32574
Super Singer Junior                       28424
(L) Formula 1 Main Race 2023              24791
FIBA: Latvia vs Lebanon                   21701
Start Music                               19582
(L) World Athletics Championships 2023    19151
Name: content, dtype: int64

## sports shows are most watched

In [49]:
distinct_contents_dictionary = content_series.to_dict()

In [78]:
# since the dictionary is too big lets only see the first 10 elements
first_few_elements = 100
for key, value in distinct_contents_dictionary.items():
  print(key, ":", distinct_contents_dictionary[key])
  first_few_elements -= 1
  if first_few_elements == 0:
    break

(L) BWF World Championships: Finals : 53302
(L) PL: Newcastle vs Liverpool : 49582
(L) PL: Sheffield vs Manchester City : 33644
The Flash [Fast Premiere] : 31410
Tiktok Awards Malaysia : 24377
Super Singer Junior : 21352
(L) Formula 1 Main Race 2023 : 18581
FIBA: Latvia vs Lebanon : 16302
Start Music : 14771
(L) World Athletics Championships 2023 : 14331
Buletin Utama : 13983
Pathaan : 13325
Varalaru : 13125
Rajuk Karina : 12911
Sepahtu Reunion Live S8 Ep07 : 12450
2000 Prime Talk : 12088
Evening Edition : 11683
Saregamapa Little Champs : 11509
Narcotics Heroes : 10115
(L) Miss Hong Kong Pageant 2023 : 9986
Baagyaletchumi Illatharisiyin Kadhai : 9374
Super Nabil Ep01 : 9129
Pandian Stores : 9095
Puss In Boots: The Last Wish : 8856
Thamizhum Saraswathiyum : 8812
Running Man : 8738
MasterChef Australia : 8729
MW3: Newcastle United vs Liverpool : 8256
(L) MSL 2023: PRK lwn KLC : 8173
Siragadikka Aasai : 7540
Aku Yang Kau Sakiti : 7336
Ethirneechal : 6766
(L) PL: Burnley vs Aston Villa : 6

# Total count of distinct TV shows i.e contents

In [79]:
len(distinct_contents_dictionary)

52133

In [52]:
datetime_format = "%Y-%m-%d %H:%M:%S.%f %Z"

# Convert the column to datetime using the specified format
sampled_df['event_start_tz'] = pd.to_datetime(sampled_df['event_start_tz'], format=datetime_format)


In [53]:
sampled_df.shape

(3725105, 6)

In [54]:
sampled_df["event_start_tz"][0]


Timestamp('2023-08-27 11:03:13.686000+0800', tz='Asia/Singapore')

In [84]:
sampled_df

state                          event_start_tz  \
14622656      Selangor  2023-08-28 19:37:11.438 Asia/Singapore   
24953294  Kuala Lumpur  2023-08-29 00:54:11.028 Asia/Singapore   
12353742      Selangor  2023-08-28 00:55:21.046 Asia/Singapore   
4993584       Selangor  2023-08-27 14:08:21.937 Asia/Singapore   
20949672         Johor  2023-08-28 13:51:03.669 Asia/Singapore   
...                ...                                     ...   
13690390  Kuala Lumpur  2023-08-28 01:48:37.945 Asia/Singapore   
9081794       Selangor  2023-08-27 20:02:38.647 Asia/Singapore   
19056004  Kuala Lumpur  2023-08-28 17:35:44.997 Asia/Singapore   
20244015      Selangor  2023-08-28 22:16:24.330 Asia/Singapore   
21151475      Selangor  2023-08-28 12:07:07.003 Asia/Singapore   

          duration_seconds content_type                         content  \
14622656               119          VOD     Lagu Didi & Friends S3 Ep31   
24953294               378          VOD             Upin & Ipin S5 Ep06   
12353742              1949       LINEAR  (L) PL: Newcastle vs Liverpool   
4993584                 87       LINEAR        A Pillow Case of Mystery   
20949672               128          VOD     Lagu Didi & Friends S2 Ep08   
...                    ...          ...                             ...   
13690390              2216          VOD      Jangan Menangis Cinta Ep18   
9081794                  0          VOD       The Flash [Fast Premiere]   
19056004                49       LINEAR  Arsenal TV : Arsenal Vs Fulham   
20244015                66          VOD         Where's Chicky? S2 Ep14   
21151475                 3   Live Pause                     Berita 7:57   

          content_source  
14622656       astro.30d  
24953294       astro.30d  
12353742        ASSP3 HD  
4993584   TVB CLASSIC HD  
20949672       astro.30d  
...                  ...  
13690390       astro.30d  
9081794     astro.com.my  
19056004     beIN SPORTS  
20244015        astro.7d  
21151475      Awesome TV  

[4966806 rows x 6 columns]

## duration_seconds is the only numerical field hence lets see description of that field

In [90]:
sampled_df.describe()

duration_seconds
count      4.966806e+06
mean       5.243420e+02
std        1.160789e+03
min        0.000000e+00
25%        1.000000e+01
50%        1.250000e+02
75%        4.270000e+02
max        3.239500e+04

In [89]:
sampled_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4966806 entries, 14622656 to 21151475
Data columns (total 6 columns):
 #   Column            Dtype 
---  ------            ----- 
 0   state             object
 1   event_start_tz    object
 2   duration_seconds  int64 
 3   content_type      object
 4   content           object
 5   content_source    object
dtypes: int64(1), object(5)
memory usage: 394.3+ MB


### ***sampled dataframe  is what % of orignal dataframe ?***

In [112]:
(sampled_df.shape[0] / main_df.shape[0]) * 100

19.0943114759449

### Now we consider only those shows whose watch duration is non-zero

In [100]:
nz_duartion_df = sampled_df[sampled_df["duration_seconds"] !=0]

In [101]:
nz_duartion_df

state                          event_start_tz  \
14622656      Selangor  2023-08-28 19:37:11.438 Asia/Singapore   
24953294  Kuala Lumpur  2023-08-29 00:54:11.028 Asia/Singapore   
12353742      Selangor  2023-08-28 00:55:21.046 Asia/Singapore   
4993584       Selangor  2023-08-27 14:08:21.937 Asia/Singapore   
20949672         Johor  2023-08-28 13:51:03.669 Asia/Singapore   
...                ...                                     ...   
21510289      Selangor  2023-08-28 21:41:16.559 Asia/Singapore   
13690390  Kuala Lumpur  2023-08-28 01:48:37.945 Asia/Singapore   
19056004  Kuala Lumpur  2023-08-28 17:35:44.997 Asia/Singapore   
20244015      Selangor  2023-08-28 22:16:24.330 Asia/Singapore   
21151475      Selangor  2023-08-28 12:07:07.003 Asia/Singapore   

          duration_seconds     content_type                         content  \
14622656               119              VOD     Lagu Didi & Friends S3 Ep31   
24953294               378              VOD             Upin & Ipin S5 Ep06   
12353742              1949           LINEAR  (L) PL: Newcastle vs Liverpool   
4993584                 87           LINEAR        A Pillow Case of Mystery   
20949672               128              VOD     Lagu Didi & Friends S2 Ep08   
...                    ...              ...                             ...   
21510289                 4  Play From Start                       Majalah 3   
13690390              2216              VOD      Jangan Menangis Cinta Ep18   
19056004                49           LINEAR  Arsenal TV : Arsenal Vs Fulham   
20244015                66              VOD         Where's Chicky? S2 Ep14   
21151475                 3       Live Pause                     Berita 7:57   

          content_source  
14622656       astro.30d  
24953294       astro.30d  
12353742        ASSP3 HD  
4993584   TVB CLASSIC HD  
20949672       astro.30d  
...                  ...  
21510289             TV3  
13690390       astro.30d  
19056004     beIN SPORTS  
20244015        astro.7d  
21151475      Awesome TV  

[4378496 rows x 6 columns]

In [109]:
nz_duartion_df.shape

(4378496, 6)

In [111]:
(nz_duartion_df.shape[0] / main_df.shape[0]) * 100

16.832621692930797

In [103]:
nz_duartion_df["state"].value_counts()

Kuala Lumpur       2439745
Selangor            927057
Penang              263188
Johor               254033
Perak               138457
Sarawak              81245
Negeri Sembilan      75902
Sabah                72932
Melaka               41072
Kedah                31033
Kelantan             21524
Pahang               18232
Terengganu           10718
Putrajaya             2525
Perlis                 706
Labuan                 127
Name: state, dtype: int64

In [106]:
sampled_df["state"].value_counts()

Kuala Lumpur       2752122
Selangor           1052497
Penang              301065
Johor               289287
Perak               161315
Sarawak              91277
Negeri Sembilan      87627
Sabah                83336
Melaka               47890
Kedah                37831
Kelantan             24785
Pahang               21128
Terengganu           12850
Putrajaya             2883
Perlis                 755
Labuan                 158
Name: state, dtype: int64

## List of distinct content providers

In [114]:
provider_list = nz_duartion_df["content_source"].value_counts()

In [118]:
provider_list

astro.30d        1428113
astro.7d          570836
astro.com.my      322057
STAR VIJAY HD     175092
TV3               143212
                  ...   
CHUTTI TV              6
beIN SPORTS 4          3
astro.7d.ads           3
ARENA BOLA 3           2
Too FUN!               1
Name: content_source, Length: 183, dtype: int64

### top 20 content providers

In [117]:
provider_list.head(20) # seeing top 20 content providers

astro.30d         1428113
astro.7d           570836
astro.com.my       322057
STAR VIJAY HD      175092
TV3                143212
astro.2d           134957
ASSP4 HD           111529
ASSP3 HD            89780
ASSP HD             78008
SUN TV HD           59203
ZEE TAMIL HD        57529
AEC HD              56604
RIA HD              44760
beIN SPORTS         38814
CERIA HD            37328
Vellithirai HD      36054
VINMEEN HD          34765
TVBJ                32027
AXN HD              31854
ONE HD              29918
Name: content_source, dtype: int64

4378496